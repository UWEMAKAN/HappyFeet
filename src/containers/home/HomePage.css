.HomePage {
  display: grid;
  column-gap: .5rem;
  grid-template-columns: 1fr 3fr;
  font-size: 1.6rem;
  background-color: #fff;
  border: 1px solid #dedede;
}

.HomePage .MainArea {
  border-left: 1px solid #dedede;
}

@media screen and (max-width: 900px) {
  .HomePage {
    column-gap: 0rem;
    grid-template-columns: 1fr 4fr;
  }
}

@media screen and (max-width: 450px) {
  .HomePage {
    grid-template-columns: 1fr;
    border: none;
  }
}
