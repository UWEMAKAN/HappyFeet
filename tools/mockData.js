module.exports = {
  shoes: [
    {
      id: 1,
      title: 'Blue Snickers',
      brand: 'Addidas',
      image: './photos/addidas-blue-1-600x600.jpg',
      color: 'Blue',
      price: 325,
      discount: 20,
      rating: 4.4,
    },
    {
      id: 2,
      title: 'Black Snickers',
      brand: 'Others',
      image: './photos/others-black-1-600x600.jpg',
      color: 'Black',
      price: 295,
      discount: 25,
      rating: 4.3,
    },
    {
      id: 3,
      title: 'Black Snickers',
      brand: 'Others',
      image: './photos/others-black-2-600x600.jpg',
      color: 'Black',
      price: 325,
      discount: 25,
      rating: 4.3,
    },
    {
      id: 4,
      title: 'White Snickers',
      brand: 'Nike',
      image: './photos/nike-white-1-600x600.jpg',
      color: 'White',
      price: 450,
      discount: 20,
      rating: 4.6,
    },
    {
      id: 5,
      title: 'Snickers',
      brand: 'Addidas',
      image: './photos/addidas-green-1-600x600.jpg',
      color: 'Green',
      price: 400,
      discount: 25,
      rating: 4.5,
    },
    {
      id: 6,
      title: 'High Hills',
      brand: 'Others',
      image: './photos/others-black-hills-1-600x600.jpg',
      color: 'Black',
      price: 375,
      discount: 25,
      rating: 4.3,
    },
    {
      id: 7,
      title: 'White Snickers',
      brand: 'Nike',
      image: './photos/nike-white-2-600x600.jpg',
      color: 'White',
      price: 450,
      discount: 20,
      rating: 4.6,
    },
    {
      id: 8,
      title: 'Blue Snickers',
      brand: 'Addidas',
      image: './photos/addidas-blue-2-600x600.jpg',
      color: 'Blue',
      price: 400,
      discount: 25,
      rating: 4.5,
    },
    {
      id: 9,
      title: 'Black Leather Shoes',
      brand: 'Others',
      image: './photos/others-black-leather-1-600x600.jpg',
      color: 'Black',
      price: 425,
      discount: 25,
      rating: 4.3,
    },
    {
      id: 10,
      title: 'White Snickers',
      brand: 'Addidas',
      image: './photos/addidas-white-1-600x600.jpeg',
      color: 'White',
      price: 400,
      discount: 25,
      rating: 4.5,
    },
    {
      id: 11,
      title: 'Blue Snickers',
      brand: 'Others',
      image: './photos/others-blue-1-600x600.jpeg',
      color: 'Blue',
      price: 420,
      discount: 25,
      rating: 4.3,
    },
  ]
};
