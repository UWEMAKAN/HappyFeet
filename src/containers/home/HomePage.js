import React, { useState } from 'react';
import propTypes from 'prop-types';
import Grid from '../../components/layout/Grid/Grid';
import SideBar from '../../components/layout/SideBar/SideBar';
import Sort from '../../components/Sort/Sort';
import { sortShoes } from './utils';
import './HomePage.css';

export const selectPriceHandler = (
  setEventFired, setSortBy, setMinPrice, setMaxPrice
) => (event) => {
  const priceSelector = event.target.name;
  const price = Number(event.target.value);
  setEventFired(true);
  setSortBy('');
  setTimeout(() => {
    if (priceSelector === 'minPriceSelector') {
      setMinPrice(price);
    }
    if (priceSelector === 'maxPriceSelector') {
      setMaxPrice(price);
    }
    setEventFired(false);
  }, 500);
};

export const searchBrandHandler = (setEventFired, setSortBy, setSearchTerm) => (event) => {
  const search = event.target.value;
  setEventFired(true);
  setSortBy('');
  setTimeout(() => {
    setSearchTerm(search);
    setEventFired(false);
  }, 500);
};

export const checkedColorsHandler = (setEventFired, setSortBy, setColors, colors) => (event) => {
  const { name } = event.target;
  setEventFired(true);
  setSortBy('');
  setTimeout(() => {
    if (colors[name]) {
      setColors({ ...colors, [name]: false });
    } else {
      setColors({ ...colors, [name]: true });
    }
    setEventFired(false);
  }, 500);
};

const HomePage = ({ shoes, buyShoe }) => {
  const [minPrice, setMinPrice] = useState(null);
  const [maxPrice, setMaxPrice] = useState(null);
  const [searchTerm, setSearchTerm] = useState('');
  const [eventFired, setEventFired] = useState(false);
  const [sortBy, setSortBy] = useState('');
  const [colors, setColors] = useState({
    blue: false,
    red: false,
    white: false,
    black: false,
    green: false,
    brown: false
  });

  const selectHandler = selectPriceHandler(setEventFired, setSortBy, setMinPrice, setMaxPrice);
  const searchHandler = searchBrandHandler(setEventFired, setSortBy, setSearchTerm);
  const checkedHandler = checkedColorsHandler(setEventFired, setSortBy, setColors, colors);

  const sortButtonsHandler = (event) => {
    const { name } = event.target;
    setSortBy(name);
  };

  const buyButtonHandler = (event) => {
    const { name } = event.target;
    const id = Number(name);
    buyShoe(id);
  };

  const shoesToRender = sortShoes(shoes, sortBy, minPrice, maxPrice, searchTerm, colors);
  return (
    <div className="HomePage">
      <SideBar
        selectHandler={selectHandler}
        searchHandler={searchHandler}
        checkedHandler={checkedHandler}
      />
      <div className="MainArea">
        <Sort sortHandler={sortButtonsHandler} />
        <Grid clickHandler={buyButtonHandler} shoes={shoesToRender} eventFired={eventFired} />
      </div>
    </div>
  );
};

HomePage.propTypes = {
  shoes: propTypes.instanceOf(Array).isRequired,
  buyShoe: propTypes.func.isRequired
};

export default HomePage;
