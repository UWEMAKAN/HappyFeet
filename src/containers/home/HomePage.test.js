/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { shallow } from 'enzyme';
import {
  cleanup, fireEvent, render, act
} from '@testing-library/react';
import HomePage, { selectPriceHandler, searchBrandHandler, checkedColorsHandler } from './HomePage';
import { shoes } from '../../../tools/mockData';

const shallowRenderHomePage = (args) => {
  const defaultProps = {
    shoes,
    buyShoe: jest.fn()
  };
  const props = { ...defaultProps, ...args };
  return shallow(<HomePage {...props} />);
};

const renderReactComponent = async (args) => {
  const defaultProps = {
    shoes,
    buyShoe: jest.fn()
  };
  const props = { ...defaultProps, ...args };
  return render(<MemoryRouter><HomePage {...props} /></MemoryRouter>);
};

describe('HomePage Component Shallow Rendered', () => {
  test('should render HomePage Component', () => {
    const wrapper = shallowRenderHomePage();
    expect.assertions(1);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('HomePage Component Mounted', () => {
  afterEach(cleanup);
  test('should fire the click event on the ascending order button', async () => {
    const { container } = await renderReactComponent();
    const name = 'ascending';
    const button = container.querySelector('button[name="ascending"]');
    act(() => {
      fireEvent.click(button, { target: { name } });
    });
    expect.assertions(1);
    expect(button).toBeTruthy();
  });

  test('should fire the click event on the descending order button', async () => {
    const { container } = await renderReactComponent();
    const name = 'descending';
    const button = container.querySelector('button[name="descending"]');
    act(() => {
      fireEvent.click(button, { target: { name } });
    });
    expect.assertions(1);
    expect(button).toBeTruthy();
  });

  test('should fire the click event on the buy button', async () => {
    const buyShoe = jest.fn();
    const { container } = await renderReactComponent({ buyShoe });
    const name = '1';
    const button = container.querySelector('button[name="1"]');
    act(() => {
      fireEvent.click(button, { target: { name } });
    });
    expect.assertions(1);
    expect(buyShoe).toHaveBeenCalledTimes(1);
  });
});

describe('Testing selectPriceHandler', () => {
  afterEach(() => {
    jest.useRealTimers();
  });

  test('should call setEventFired, setSortBy, setMinPrice', () => {
    jest.useFakeTimers();
    const setEventFired = jest.fn();
    const setSortBy = jest.fn();
    const setMinPrice = jest.fn();
    const setMaxPrice = jest.fn();
    const name = 'minPriceSelector';
    const value = '2000';
    const event = { target: { value, name } };
    const selectHandler = selectPriceHandler(setEventFired, setSortBy, setMinPrice, setMaxPrice);
    selectHandler(event);
    jest.runAllTimers();
    expect.assertions(4);
    expect(setEventFired).toHaveBeenCalledTimes(2);
    expect(setSortBy).toHaveBeenCalledTimes(1);
    expect(setMinPrice).toHaveBeenCalledTimes(1);
    expect(setMaxPrice).toHaveBeenCalledTimes(0);
  });

  test('should call setEventFired, setSortBy, setMaxPrice', () => {
    jest.useFakeTimers();
    const setEventFired = jest.fn();
    const setSortBy = jest.fn();
    const setMinPrice = jest.fn();
    const setMaxPrice = jest.fn();
    const name = 'maxPriceSelector';
    const value = '3000';
    const event = { target: { value, name } };
    const selectHandler = selectPriceHandler(setEventFired, setSortBy, setMinPrice, setMaxPrice);
    selectHandler(event);
    jest.runAllTimers();
    expect.assertions(4);
    expect(setEventFired).toHaveBeenCalledTimes(2);
    expect(setSortBy).toHaveBeenCalledTimes(1);
    expect(setMinPrice).toHaveBeenCalledTimes(0);
    expect(setMaxPrice).toHaveBeenCalledTimes(1);
  });
});

describe('Testing searchBrandHandler', () => {
  afterEach(() => {
    jest.useRealTimers();
  });

  test('should call setEventFired, setSortBy, setSearchTerm', () => {
    jest.useFakeTimers();
    const setEventFired = jest.fn();
    const setSortBy = jest.fn();
    const setSearchTerm = jest.fn();
    const name = 'search';
    const value = 'ni';
    const event = { target: { value, name } };
    const searchHandler = searchBrandHandler(setEventFired, setSortBy, setSearchTerm);
    searchHandler(event);
    jest.runAllTimers();
    expect.assertions(3);
    expect(setEventFired).toHaveBeenCalledTimes(2);
    expect(setSortBy).toHaveBeenCalledTimes(1);
    expect(setSearchTerm).toHaveBeenCalledTimes(1);
  });
});

describe('Testing checkedColorHandler', () => {
  afterEach(() => {
    jest.useRealTimers();
  });

  test('should call setEventFired, setSortBy, setColors', () => {
    jest.useFakeTimers();
    const colors = {
      blue: false,
      red: false,
      white: false,
      black: false,
      green: false,
      brown: false
    };
    const setEventFired = jest.fn();
    const setSortBy = jest.fn();
    const setColors = jest.fn();
    const name = 'black';
    const event = { target: { name } };
    const checkedHandler = checkedColorsHandler(setEventFired, setSortBy, setColors, colors);
    checkedHandler(event);
    jest.runAllTimers();
    expect.assertions(3);
    expect(setEventFired).toHaveBeenCalledTimes(2);
    expect(setSortBy).toHaveBeenCalledTimes(1);
    expect(setColors).toHaveBeenCalledTimes(1);
  });

  test('should call setEventFired, setSortBy, setColors', () => {
    jest.useFakeTimers();
    const colors = {
      blue: true,
      red: false,
      white: false,
      black: false,
      green: false,
      brown: false
    };
    const setEventFired = jest.fn();
    const setSortBy = jest.fn();
    const setColors = jest.fn();
    const name = 'blue';
    const event = { target: { name } };
    const checkedHandler = checkedColorsHandler(setEventFired, setSortBy, setColors, colors);
    checkedHandler(event);
    jest.runAllTimers();
    expect.assertions(3);
    expect(setEventFired).toHaveBeenCalledTimes(2);
    expect(setSortBy).toHaveBeenCalledTimes(1);
    expect(setColors).toHaveBeenCalledTimes(1);
  });
});
