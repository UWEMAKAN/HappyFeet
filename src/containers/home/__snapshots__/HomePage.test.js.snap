// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage Component Shallow Rendered should render HomePage Component 1`] = `ShallowWrapper {}`;
